include/ginkgo/core/matrix/batch_dense.hpp

```diff
@@ -0,0 +1,368 @@
+     *
+     * @return the pointer to the array of values
+     */
+    value_type* get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @copydoc get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type* get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * Returns a single element for a particular batch item.
+     *
+     * @param batch_id  the batch item index to be queried
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU Dense object
+     *        from the OMP may result in incorrect behaviour)
+     */
+    value_type& at(size_type batch_id, size_type row, size_type col)
+    {
+        GKO_ASSERT(batch_id < this->get_num_batch_items());
+        return values_.get_data()[linearize_index(batch_id, row, col)];
+    }
+
+    /**
+     * @copydoc Dense::at(size_type, size_type, size_type)
+     */
+    value_type at(size_type batch_id, size_type row, size_type col) const
+    {
+        GKO_ASSERT(batch_id < this->get_num_batch_items());
+        return values_.get_const_data()[linearize_index(batch_id, row, col)];
+    }
+
+    /**
+     * Returns a single element for a particular batch item.
+     *
+     * Useful for iterating across all elements of the matrix.
+     * However, it is less efficient than the two-parameter variant of this
+     * method.
+     *
+     * @param batch_id  the batch item index to be queried
+     * @param idx  a linear index of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU Dense object
+     *        from the OMP may result in incorrect behaviour)
+     */
+    ValueType& at(size_type batch_id, size_type idx) noexcept
+    {
+        return values_.get_data()[linearize_index(batch_id, idx)];
+    }
+
+    /**
+     * @copydoc Dense::at(size_type, size_type, size_type)
+     */
+    ValueType at(size_type batch_id, size_type idx) const noexcept
+    {
+        return values_.get_const_data()[linearize_index(batch_id, idx)];
+    }
+
+    /**
+     * Returns a pointer to the array of values of the matrix for a
+     * specific batch item.
+     *
+     * @param batch_id  the id of the batch item.
+     *
+     * @return the pointer to the array of values
+     */
+    value_type* get_values_for_item(size_type batch_id) noexcept
+    {
+        GKO_ASSERT(batch_id < this->get_num_batch_items());
+        return values_.get_data() +
+               this->get_size().get_cumulative_offset(batch_id);
```

Comment:
I still think the batch_dim is not a good place for get_cumulative_offset especially when we add the stride in the future.

Reply:
Removed from batch_dim now and added to the classes themselves.